src/tribler/core/utilities/process_manager/tests/test_manager.py

```diff
@@ -0,0 +1,151 @@
+        # believes the running process with the same PID is a new process, different from the process in the DB
+        connection.execute('update processes set started_at = started_at - (60 * 60 * 24 * 100) where "primary" = 1')
+
+    p3 = TriblerProcess.current_process(ProcessKind.Core, manager=process_manager)
+    p3.pid += 2
+    # Now p3 can become a new primary process, because the previous primary process considered
+    # already finished and replaced with a new unrelated process with the same PID
+    assert p3.become_primary()
+    assert p3.primary
+
+    with process_manager.connect() as connection:
+        rows = connection.execute('select rowid from processes where "primary" = 1').fetchall()
+        # At the end, the DB should contain only one primary process, namely p3
+        assert len(rows) == 1 and rows[0][0] == p3.rowid
+
+
+def test_save(process_manager: ProcessManager):
+    p = TriblerProcess.current_process(ProcessKind.Core, manager=process_manager)
+    p.pid = p.pid + 100
+    p.save()
+    assert p.rowid is not None
+
+
+def test_set_api_port(process_manager: ProcessManager):
+    process_manager.current_process.set_api_port(12345)
+    assert process_manager.current_process.api_port == 12345
+    with process_manager.connect() as connection:
+        rows = connection.execute('select rowid from processes where api_port = 12345').fetchall()
+        assert len(rows) == 1 and rows[0][0] == process_manager.current_process.rowid
+
+
+@patch('sys.exit')
+def test_sys_exit(sys_exit: Mock, process_manager: ProcessManager):
+    process_manager.sys_exit(123, 'Error text')
+
+    with process_manager.connect() as connection:
+        rows = connection.execute('select "primary", error_msg from processes where rowid = ?',
+                                  [process_manager.current_process.rowid]).fetchall()
+        assert len(rows) == 1 and rows[0] == (0, 'Error text')
+    assert sys_exit.called and sys_exit.call_args[0][0] == 123
+
+
+def test_get_last_processes(process_manager: ProcessManager):
+    last_processes = process_manager.get_last_processes()
+    assert len(last_processes) == 1 and last_processes[0].rowid == process_manager.current_process.rowid
+
+    fake_process = TriblerProcess.current_process(ProcessKind.Core, manager=process_manager)
+    fake_process.pid = fake_process.pid + 1
+    fake_process.become_primary()
+
+    last_processes = process_manager.get_last_processes()
+    assert len(last_processes) == 2
+    assert last_processes[0].rowid == process_manager.current_process.rowid
+    assert last_processes[1].rowid == fake_process.rowid
+
+
+@patch.object(logger, 'warning')
+def test_global_process_manager(warning: Mock, process_manager: ProcessManager):
+    assert get_global_process_manager() is None
+
+    set_global_process_manager(process_manager)
+    assert get_global_process_manager() is process_manager
+
+    set_global_process_manager(None)
+    assert get_global_process_manager() is None
```

Comment:
This test uses a singleton for storing the global process manager. This could lead to a flaky behavior for the test suite as we had before.

The test as well not tests the main functionality of setter (the locking).

That's why it is better to just remove it.

Reply:
I'm ok with removing the test, but I don't see the reason for it, so I suggest discussing it one more time. In my opinion, it is better to have this test for better coverage of the code that handles the global singleton. Without the coverage, it may be possible that the logic will be broken in some later refactoring.

The test does not contain any complex logic and returns the global singleton to a previous state right after the test. 

> The test as well not tests the main functionality of setter (the locking).

I don't think the lock is the main functionality of `get_global_process_manager`/`set_global_process_manager`. There are two types of locks - long and short. If the code contains a non-trivial code inside a long lock, it may spend too much time inside it, and another thread can freeze for too long to acquire it. An even worse scenario is a deadlock when two threads acquire two different locks in an improper order.

Neither of these potential problems is possible in the current test. The logic inside the lock is very simple and contains just a single Python statement:
```python
    with _lock:
        global_process_manager = process_manager
```
and
```
    with _lock:
        return global_process_manager
```
It is an example of a short lock that only covers a simple instruction to protect access to a global variable to make read/write operation atomic. It is just not possible to stay inside the lock for too long. There is no non-trivial function call inside the lock that can delay the exit from the lock. So, in this case, the lock is not a "main" functionality of the function; it is a trivial and very short lock for protecting the atomic access to a global variable. The lock functionality is already tested in Python's test suite, and we can quickly check the code inside the lock to see that it does not have any potential to create a problem.

Reply:
Update: I added a try/finally block inside the test to ensure it has no chance to affect the following tests in case of an exception. In any case, the test suite will fail in case of exception, so there is no chance that after this test, we can have a mysterious problem in a subsequent test, but this test passes without an error. So, even in the (highly improbable) scenario that this test will fail and affect the global state, it will also be marked as failed and should explain the actual reason for any other error.

Reply:
Why don't you use simple singleton creation for the `ProcessManager` as we have done for the `ExceptionHandler`?

https://github.com/Tribler/tribler/blob/main/src/tribler/core/components/reporter/exception_handler.py#L131 

In this case, it is not necessary to test the setter and the getter.

Reply:
> I don't think the lock is the main functionality of get_global_process_manager/set_global_process_manager.

Why then do use locks? What corner case do you want to cover?

Reply:
>  I need to create TriblerProcess before it, and I prefer to create the TriblerProcess instance in tribler.gui.start_gui/tribler.core.start_core instead.

It is not necessary. You can create a global singleton before and set the current process to it after.

Reply:
> Why then do use locks? What corner case do you want to cover?

It is necessary to have a lock when it is potentially possible to work with the same global variable from two different threads

> It is not necessary. You can create a global singleton before and set the current process to it after.

The the API will be more complicated, as it becomes necessary to make the current_process optional and check it in all places

Ok, I do not agree with this, but I removed the test to avoid long disputes